bigbite/themer
# [BBT-87] Debounce preview CSS updates

## Description

This PR aims limits how and when the remote API is called to fetch updated CSS data for rendering the preview.

Fixes [BBT-87](https://b5ecom.atlassian.net/browse/BBT-87)

This is achieved in two ways:
- Implementing `debounce` to limit the number of 
- Implementing an `AbortController` to cancel any in-flight API requests when react re-renders.

This functionality is extracted into a custom hook named `useDebouncedApiFetch`.

We currently debounce the requests with a very low value, which ensures that the preview still responds to user changes quickly whilst conducting significantly fewer calls to the API than previously.

The `AbortController` ensures that if duplicate/concurrent requests are made, that all but the most recent will be cancelled.

## Change Log
- Implement `useDebouncedApiFetch` hook
- Debounce preview CSS API calls by 100ms

## Steps to test

You can inspect the difference this change makes in the network tab. Ensure that any changes to the theme UI result in the preview CSS being updated as previously, but notice that significantly fewer calls are made to the API.

Calls cancelled by the AbortController will appear in red.

## Screenshots/Videos

#### Before:

https://github.com/bigbite/themer/assets/41474928/ec0761ee-2683-4ba1-8d50-39012a94047c



#### After:

https://github.com/bigbite/themer/assets/41474928/0260400c-f42b-4b2a-b5bb-d937ba13cf66


## Checklist:





[BBT-87]: https://b5ecom.atlassian.net/browse/BBT-87?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
